StorageOS Volume Plugin

croomes opened the pull request:
**What this PR does / why we need it**:
This PR adds a new volume plugin for StorageOS volumes.  StorageOS runs as a container on Kubelet nodes, aggregating local or attached storage and making its capacity available to all nodes within the cluster. More information at http://storageos.com.

The StorageOS plugin supports:

1. Dynamic Provisioning using Storage Classes
2. Persistent Volumes and Persistent Volume Claims.

**Which issue this PR fixes**
A feature request has been created:
https://github.com/kubernetes/features/issues/190

This isn't on the schedule for 1.6 as I wasn't sure when it would be ready.  We intend to make the StorageOS container openly available within the 1.6 lifetime.

**Special notes for your reviewer**:

Separate commits for feature and godep changes.

**Release note**:
```release-note
StorageOS Volume Driver
[StorageOS](http://www.storageos.com) can be used as a storage provider for Kubernetes.  With StorageOS, capacity from local or attached storage is pooled across the cluster, providing converged infrastructure for cloud-native applications. 
```

croomes committed:
Add StorageOS volume plugin

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/42156)
<!-- Reviewable:end -->

croomes commented:
I'll rebase this as soon as I get back from DockerCon

rootfs commented:
/assign

rootfs commented:
@k8s-bot ok to test

rootfs commented on pkg/api/types.go:
```diff
@@ -1110,6 +1116,33 @@ type ScaleIOVolumeSource struct {
+	// namespace is specified, the namespace of the pod will be used.  Set to
+	// "default" if you are not using namespaces within StorageOS.
+	// +optional
+	Namespace string
```
why do you need NameSpace in volume source?

croomes replied:
We use namespaces within StorageOS too, so by default it will use the Pod's namespace but you can force it not to use StorageOS namespace's at all by setting to "default".  You can also specify a different StorageOS namespace here.  Normally the StorageOS namespaces will mirror Kubernetes'.

rootfs commented on pkg/volume/storageos/doc.go:
```diff
@@ -0,0 +1,19 @@
+/*
+Copyright 2015 The Kubernetes Authors.
```
2017

croomes replied:
done

rootfs commented on pkg/volume/storageos/storageos_test.go:
```diff
@@ -0,0 +1,333 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
2017

croomes replied:
done

rootfs commented on pkg/volume/storageos/storageos.go:
```diff
@@ -0,0 +1,564 @@
+			podUID:          pod.UID,
+			pvName:          spec.Name(),
+			volName:         source.VolumeName,
+			namespace:       source.Namespace,
```
you can get namespace from Pod `pod.Namespace`

croomes replied:
As comment above, this is the volume namespace and could be different from the Pod namespace.

rootfs commented on pkg/volume/storageos/storageos.go:
```diff
@@ -0,0 +1,564 @@
+		storageos: &storageos{
+			pvName:    spec.Name(),
+			volName:   spec.PersistentVolume.Spec.StorageOS.VolumeName,
+			namespace: spec.PersistentVolume.Spec.StorageOS.Namespace,
```
you can get namespace from PV's ClaimRef

croomes replied:
This is the volume's namespace, not necessarily the Pod's.  Starting to think I should rename it volNamespace to make it more clear - what do you think?

rootfs commented on pkg/volume/storageos/storageos.go:
```diff
@@ -0,0 +1,564 @@
+
+	// Set parameters from StorageClass
+	if len(v.options.Parameters) > 0 {
+		if description, ok := v.options.Parameters["description"]; ok {
```
iterate parameters and use switch/case

croomes replied:
much better - done

rootfs commented on pkg/volume/storageos/storageos.go:
```diff
@@ -0,0 +1,564 @@
+	v.volName = v.options.PVName
+	v.namespace = v.options.PVC.Namespace
+	v.labels = make(map[string]string)
+	for k, val := range v.options.PVC.Labels {
```
why setting labels here?

croomes replied:
We copy the labels from the PVC to the requested volume.  StorageOS uses the labels from k8s to set policy.  Fox example, if the PVC has environment=production set, a policy rule within StorageOS might set the number of replicas to 2.

rootfs commented on pkg/volume/storageos/storageos.go:
```diff
@@ -0,0 +1,564 @@
+			pv.Labels = make(map[string]string)
+		}
+		for k, v := range vol.Labels {
+			pv.Labels[k] = v
```
passing labels from PVC to PV?

croomes replied:
This is setting labels created by StorageOS when the volume is provisioned and applying them to the PV.  So in the example above, the label `storageos.feature.replicas=2` would be set on the PV.  These labels are only informational.

croomes commented:
Hi @rootfs - thanks for the review.  I'm updating and will reply to the remaining points soon.

croomes commented:
@rootfs I'm not sure I understand the test failures, do you think these are caused by my commit or unrelated?  Otherwise do my comments/fixes above make sense or need more work?

rootfs commented:
@croomes please run `hack/update-godep-licenses.sh`

croomes commented:
pull-kubernetes-cross failure looks unrelated: https://github.com/kubernetes/kubernetes/issues/45554

@k8s-bot kops aws e2e test this

croomes commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

croomes commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

croomes commented:
@k8s-bot pull-kubernetes-verify test this

croomes commented:
@k8s-bot pull-kubernetes-node-e2e test this

croomes commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

croomes commented:
@k8s-bot pull-kubernetes-unit test this

croomes commented:
@k8s-bot pull-kubernetes-cross test this
@k8s-bot pull-kubernetes-kubemark-e2e-gce test this

liggitt commented on pkg/api/v1/types.go:
```diff
@@ -1220,6 +1226,33 @@ type ScaleIOVolumeSource struct {
+	// namespace is specified, the namespace of the pod will be used.  Set to
+	// "default" if you are not using namespaces within StorageOS.
+	// +optional
+	Namespace string `json:"namespace,omitempty" protobuf:"bytes,2,opt,name=namespace"`
```
Is there something else we can call this? I'm concerned it will be confused for a kubernetes namespace and used incorrectly

croomes replied:
agreed - renamed to VolumeNamespace.

Appreciate the review.

croomes commented:
@k8s-bot pull-kubernetes-cross test this

croomes commented:
The kubernetes-cross fail looks unrelated and failing across most PRs since this morning.

@saad-ali I'm afk with no internet until Thursday lunchtime (GMT) so I've handed over to my colleagues @rusenask and @SpectralHiss.  Really hoping this can make 1.7.  Thanks!

rusenask commented:
updated PR since we saw some conflicts after local plugin was added

rusenask commented:
@k8s-bot pull-kubernetes-verify test this

rusenask commented:
@k8s-bot pull-kubernetes-verify test this

msau42 commented on examples/persistent-volume-provisioning/README.md:
````diff
@@ -94,6 +94,26 @@ parameters:
+
+```yaml
+kind: StorageClass
+apiVersion: storage.k8s.io/v1beta1
````
We have storage.k8s.io/v1 now

msau42 commented on examples/volumes/storageos/README.md:
```diff
@@ -0,0 +1,466 @@
+   metadata:
+     name: fast0001
+     annotations:
+       volume.beta.kubernetes.io/storage-class: fast
```
storageClassName is a field in the spec now

rusenask commented:
@rootfs tests are passing again after github recovered. Is there anything else we need to do to push this forward? not sure what's the process from our side as I wasn't involved in this and Simon is still away.

msau42 commented on pkg/volume/storageos/storageos.go:
```diff
@@ -0,0 +1,574 @@
+		volName = volumeName
+	}
+	storageosVolume := &v1.Volume{
+		Name: volumeName,
```
I'm not sure if this is prone to the volume reconstruction issue. @chakri-nelluri, can you check?

msau42 reviewed (commented):
I just had some minor comments.  They can be addressed in a separate PR.

rootfs commented:
/assign @msau42 

@croomes please contact @thockin for approval.

croomes commented:
@k8s-bot pull-kubernetes-cross test this

croomes commented:
@k8s-bot pull-kubernetes-cross test this

msau42 commented:
/lgtm

falenn commented:
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  curtis.l.bates@gmail.com
    Domain imwiz.com has exceeded the max emails per hour (166/150 (110%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
------ The body of the message is 5374 characters long; only the first
------ 5000 or so are included here.
Received: from github-smtp2-ext1.iad.github.net ([192.30.252.192]:54544 helo=github-smtp2a-ext-cp1-prd.iad.github.net)
	by box969.bluehost.com with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.87)
	(envelope-from <noreply@github.com>)
	id 1dGU7o-000QBh-Qm
	for dev@imwiz.com; Thu, 01 Jun 2017 11:43:09 -0600
Date: Thu, 01 Jun 2017 10:42:58 -0700
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1496338978;
	bh=svPPSpHFJ2MUIr/MGfv97CwEwwbNYcPXlzzsv8X81vw=;
	h=From:Reply-To:To:Cc:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=CbZ6fqOk5aVdriQ9bX3nuNhogTKjc9qdZtnkK7q3NxpNsubJlZDqNgB4hNHj0HC3g
	 dRjQrK5hhc1THAt2XPbeGREcV/BdkOKQKgvJMQKB/esiFPBiSD7yfRQqkvuehywyep
	 tsdQ4j8NaEZWsvrtxkE9vnYnBwZwjYiU0JHKvzKw=
From: Kubernetes Submit Queue <notifications@github.com>
Reply-To: kubernetes/kubernetes <reply@reply.github.com>
To: kubernetes/kubernetes <kubernetes@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Message-ID: <kubernetes/kubernetes/pull/42156/c305567207@github.com>
In-Reply-To: <kubernetes/kubernetes/pull/42156@github.com>
References: <kubernetes/kubernetes/pull/42156@github.com>
Subject: Re: [kubernetes/kubernetes] StorageOS Volume Plugin (#42156)
Mime-Version: 1.0
Content-Type: multipart/alternative;
 boundary="--==_mimepart_593052221ac0_7f803ffb1f6c5c3c53285";
 charset=UTF-8
Content-Transfer-Encoding: 7bit
Precedence: list
X-GitHub-Sender: k8s-merge-robot
X-GitHub-Recipient: falenn
X-GitHub-Reason: subscribed
List-ID: kubernetes/kubernetes <kubernetes.kubernetes.github.com>
List-Archive: https://github.com/kubernetes/kubernetes
List-Post: <mailto:reply@reply.github.com>
List-Unsubscribe: <mailto:unsub+000ab60a4dd903874a395b4da5585758186fc7fd65bb93c892cf000000011548142292a169ce0c8b9ea8@reply.github.com>,
 <https://github.com/notifications/unsubscribe/AAq2Cqd9C7aK9FXnROIJslWyH2Idz5nnks5r_vgigaJpZM4MNExi>
X-Auto-Response-Suppress: All
X-GitHub-Recipient-Address: dev@imwiz.com
X-Spam-Status: No, score=-4.0
X-Spam-Score: -39
X-Spam-Bar: ----
X-Ham-Report: Spam detection software, running on the system "box969.bluehost.com",
 has NOT identified this incoming email as spam.  The original
 message has been attached to this so you can view it or label
 similar future email.  If you have any questions, see
 root\@localhost for details.
 
 Content preview:  [APPROVALNOTIFIER] This PR is **NOT APPROVED** This pull-request
    has been approved by: *<a href="https://github.com/kubernetes/kubernetes/pull/42156#"
    title="Author self-approved">croomes</a>*, *<a href="https://github.com/kubernetes/kubernetes/pull/42156#issuecomment-305565030"
    title="LGTM">msau42</a>* We suggest the following additional approver: **jbeda**
    [...] 
 
 Content analysis details:   (-4.0 points, 4.0 required)
 
  pts rule name              description
 ---- ---------------------- --------------------------------------------------
  0.5 RCVD_IN_SORBS_SPAM     RBL: SORBS: sender is a spam source
                             [192.30.252.192 listed in dnsbl.sorbs.net]
 -5.0 RCVD_IN_DNSWL_HI       RBL: Sender listed at http://www.dnswl.org/, high
                             trust
                             [192.30.252.192 listed in list.dnswl.org]
 -0.0 SPF_PASS               SPF: sender matches SPF record
 -0.0 RCVD_IN_MSPIKE_H3      RBL: Good reputation (+3)
                             [192.30.252.192 listed in wl.mailspike.net]
  0.0 HTML_MESSAGE           BODY: HTML included in message
 -0.1 DKIM_VALID_AU          Message has a valid DKIM or DK signature from author's
                             domain
 -0.1 DKIM_VALID             Message has at least one valid DKIM or DK signature
  0.1 DKIM_SIGNED            Message has a DKIM or DK signature, not necessarily valid
 -0.0 RCVD_IN_MSPIKE_WL      Mailspike good senders
  0.7 AWL                    AWL: Adjusted score from AWL reputation of From: address
X-Spam-Flag: NO


----==_mimepart_593052221ac0_7f803ffb1f6c5c3c53285
Content-Type: text/plain;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/kubernetes/kubernetes/pull/42156#" title="Author self-approved">croomes</a>*, *<a href="https://github.com/kubernetes/kubernetes/pull/42156#issuecomment-305565030" title="LGTM">msau42</a>*
We suggest the following additional approver: **jbeda**

Assign the PR to them by writing `/assign @jbeda` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).

<details open>
Needs approval from an approver in each of these OWNERS Files:

- **[OWNERS](https://github.com/kubernetes/kubernetes/blob/master/OWNERS)**

You can indicate your approval by writing `/approve` in a comment
You can cancel your approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["jbeda"]} -->

-- 
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub:
https://github.com/kubernetes/kubernetes/pull/42156#issuecomment-305567207
----==_mimepart_593052221ac0_7f803ffb1f6c5c3c53285
Content-Type: text/html;
 charset=UTF-8
Content-Transfer-Encoding: quoted-printable

<p>[APPROVALNOTIFIER] This PR is <strong>NOT APPROVED</strong></p>
<p>This pull-request has been approved by: <em><a href=3D"https://github.=
com/kubernetes/kubernetes/pull/42156#">croomes</a></em>, <em><a href=3D"h=
ttps://github.com/kubernetes/kubernetes/pull/42156#issuecomment-305565030=
">msau42</a></em><br>
We suggest the following additional approver: <strong>jbeda</strong></p>
<p>Assign the PR to them by writing <code>/assign @jbeda</code> in a comm=
ent when ready.</p>
<p>The full list of commands accepted by this bot can be found <a href=3D=
"https://github.com/kubernetes/test-infra/blob/master/commands.md">here</=
a>.</p>
<details open=3D"">
Needs approval from an approver in each of these OWNERS Files:
<ul>
<li><strong><a href=3D"https://github.com/kubernetes/kubernetes/blob/mast=
er/OWNERS">OWNERS</a></strong></li>
</ul>
<p>You can indicate your approval by writing <code>/approve</code> in a c=
omment<br>
You can cancel your approval by writing <code>/approve cancel</code> in a=
 comment</p>
</details>


<p style=3D"font-size:small;-webkit-text-size-adjust:none;color:#666;">&m=
dash;<br />You are receiving this because you are subscribed to this thre=
ad.<br />Reply to this email directly, <a href=3D"https://github.com/kube=
rnetes/kubernetes/pull/42156#issuecomment-305567207">view it on GitHub</a=
>, or <a href=3D"https://github.com/notifications/unsubscribe-auth/AAq2Cq=
DzmA-OWTX1sNeo5DCdMX3vc5jeks5r_vgigaJpZM4MNExi">mute the thread</a>.<img =
alt=3D"" height=3D"1" src=3D"https://github.com/notifications/beacon/AAq2=
Cjipg2GWLCOsid6HEsyYDWWzcv9Aks5r_vgigaJpZM4MNExi.gif" width=3D"1" /></p>
<div itemscope itemtype=3D"http://schema.org/EmailMessage">
<div itemprop=3D"action" itemscope itemtype=3D"http://schema.org/ViewActi=
on">
  <link itemprop=3D"url" href=3D"https://github.com/kubernetes/kubernetes=
/pull/42156#issuecomment-305567207"></link>
  <meta itemprop=3D"name" content=3D"View Pull Request"></meta>
</div>
<meta itemprop=3D"description" content=3D"View this Pull Request on GitHu=
b"></meta>
</div>

<script type=3D"application/json" data-scope=3D"inboxmarkup">{"api_versio=
n":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name"=
:"GitHub"},"entity":{"external_key":"github/kubernetes/kubernetes","title=
":"kubernetes/kubernetes","subtitle":"GitHub repository","main_image_url"=
:"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d8=
8-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubuse=
rcontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.=
png","action":{"name":"Open in GitHub","url":"https://github.com/kubernet=
es/kubernetes"}},"updates":{"snippets":[{"icon":"PERSON","message":"@k8s-=
merge-robot in #42156: [APPROVALNOTIFIER] This PR is **NOT APPROVED**\n\n=
This pull-request has been approved by: *\u003ca href=3D\"https://github.=
com/kubernetes/kubernetes/pull/42156#\" title=3D\"Author self-approved\"\=
u003ecroomes\u003c/a\u003e*, *\u003ca href=3D\"https://github.com/kuberne=
tes/kubernetes/pull/42156#issuecomment-305565030\" title=3D\"LGTM\"\u003e=
msau42\u003c/a\u003e*\nWe suggest the following additional approver: **jb=
eda**\n\nAssign the PR to them by writing `/assign @jbeda` in a comment w=
hen ready.\n\nThe full list of commands accepted by this bot can be found=
 [here](https://github.com/kubernetes/test-infra/blob/master/commands.md)=
.\n\n\u003cdetails open\u003e\nNeeds approval from an approver in each of=
 these OWNERS Files:\n\n- **[OWNERS](https://github.com/kubernetes/kubern=
etes/blob/master/OWNERS)**\n\nYou can indicate your approval by writing

falenn commented:
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  curtis.l.bates@gmail.com
    Domain imwiz.com has exceeded the max emails per hour (174/150 (115%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from o9.sgmail.github.com ([167.89.101.2]:48977)
	by box969.bluehost.com with esmtps (TLSv1.2:ECDHE-RSA-AES128-GCM-SHA256:128)
	(Exim 4.87)
	(envelope-from <bounces+848413-5c7e-dev=imwiz.com@sgmail.github.com>)
	id 1dGU8F-000QQE-Au
	for dev@imwiz.com; Thu, 01 Jun 2017 11:43:35 -0600
DKIM-Signature: v=1; a=rsa-sha1; c=relaxed/relaxed; d=github.com; 
	h=from:reply-to:to:cc:in-reply-to:references:subject:mime-version:content-type:content-transfer-encoding:list-id:list-archive:list-post:list-unsubscribe; 
	s=s20150108; bh=jEE2F33KECwH2IY0aNodhjBmzkY=; b=D5DyiNVgIPAwO3z6
	jXaS7Wm71EnsRXitMPrhm90ca3leJpw1lL7ceoeS0eeE+1VTmL428wc7KJGs9WfF
	VbQhXdeXhJXtGmW6bpATi8VAppGPhqVlwAJa9MKXf9E9dRnmS20R7hNWkDYi/fIY
	i7iRWWDQlSKU3vln37DvGBgzKmQ=
Received: by filter0828p1mdw1.sendgrid.net with SMTP id filter0828p1mdw1-19891-59305228-20
        2017-06-01 17:43:04.438692086 +0000 UTC
Received: from github-smtp2b-ext-cp1-prd.iad.github.net (github-smtp2b-ext-cp1-prd.iad.github.net [192.30.253.17])
	by ismtpd0002p1iad1.sendgrid.net (SG) with ESMTP id fUjMH0AoRiWKEh6FPTzzyw
	for <dev@imwiz.com>; Thu, 01 Jun 2017 17:43:04.405 +0000 (UTC)
Date: Thu, 01 Jun 2017 10:43:04 -0700
From: Kubernetes Submit Queue <notifications@github.com>
Reply-To: kubernetes/kubernetes <reply@reply.github.com>
To: kubernetes/kubernetes <kubernetes@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Message-ID: <kubernetes/kubernetes/pull/42156/c305567214@github.com>
In-Reply-To: <kubernetes/kubernetes/pull/42156@github.com>
References: <kubernetes/kubernetes/pull/42156@github.com>
Subject: Re: [kubernetes/kubernetes] StorageOS Volume Plugin (#42156)
Mime-Version: 1.0
Content-Type: multipart/alternative;
 boundary="--==_mimepart_593052288ca2_14113f9bb3f3dc3c1098a9";
 charset=UTF-8
Content-Transfer-Encoding: 7bit
Precedence: list
X-GitHub-Sender: k8s-merge-robot
X-GitHub-Recipient: falenn
X-GitHub-Reason: subscribed
List-ID: kubernetes/kubernetes <kubernetes.kubernetes.github.com>
List-Archive: https://github.com/kubernetes/kubernetes
List-Post: <mailto:reply@reply.github.com>
List-Unsubscribe: <mailto:unsub+000ab60ae6fffef5ae2b9efe43fe94070f1e40855944300292cf000000011548142892a169ce0c8b9ea8@reply.github.com>,
 <https://github.com/notifications/unsubscribe/AAq2CjAqbUrMLAQKUMMwFSy8zvSmgQhJks5r_vgogaJpZM4MNExi>
X-Auto-Response-Suppress: All
X-GitHub-Recipient-Address: dev@imwiz.com
X-SG-EID: APO41b8ovafPb3SK9rw3vPPmuTd8rBMPcbGiCJEzxCh03EBDg9ii0jOVgvlFaJpAr8ZYV8gzhcBvAp
 6cw3IKyq6S4rkjjZxu1SaKDfSYuImDxLUxYTAaoPFPo6Pnhmk5pMwl4IUFy8mKZU8HCEEsuYv1D0de
 aw0aGVgc8nyi9cWuDumTTwhaS/dg29zcZBr3rOLpQpEaFCLpsaeG3vl32yQxMKMuK3KBXXWNtUylRt
 8=
X-Spam-Status: No, score=-2.8
X-Spam-Score: -27
X-Spam-Bar: --
X-Ham-Report: Spam detection software, running on the system "box969.bluehost.com",
 has NOT identified this incoming email as spam.  The original
 message has been attached to this so you can view it or label
 similar future email.  If you have any questions, see
 root\@localhost for details.
 
 Content preview:  @croomes PR needs rebase -- You are receiving this because
    you are subscribed to this thread. Reply to this email directly or view it
    on GitHub: https://github.com/kubernetes/kubernetes/pull/42156#issuecomment-305567214
    [...] 
 
 Content analysis details:   (-2.8 points, 4.0 required)
 
  pts rule name              description
 ---- ---------------------- --------------------------------------------------
 -0.0 RP_MATCHES_RCVD        Envelope sender domain matches handover relay domain
 -0.0 SPF_PASS               SPF: sender matches SPF record
  0.0 HTML_MESSAGE           BODY: HTML included in message
  0.7 HTML_IMAGE_ONLY_20     BODY: HTML: images with 1600-2000 bytes of words
 -0.1 DKIM_VALID_AU          Message has a valid DKIM or DK signature from author's
                             domain
 -0.1 DKIM_VALID             Message has at least one valid DKIM or DK signature
  0.1 DKIM_SIGNED            Message has a DKIM or DK signature, not necessarily valid
 -2.8 RCVD_IN_MSPIKE_H2      RBL: Average reputation (+2)
                             [167.89.101.2 listed in wl.mailspike.net]
 -0.6 AWL                    AWL: Adjusted score from AWL reputation of From: address
X-Spam-Flag: NO

----==_mimepart_593052288ca2_14113f9bb3f3dc3c1098a9
Content-Type: text/plain;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

@croomes PR needs rebase

-- 
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub:
https://github.com/kubernetes/kubernetes/pull/42156#issuecomment-305567214
----==_mimepart_593052288ca2_14113f9bb3f3dc3c1098a9
Content-Type: text/html;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

<p><a href="https://github.com/croomes" class="user-mention">@croomes</a> PR needs rebase</p>

<p style="font-size:small;-webkit-text-size-adjust:none;color:#666;">&mdash;<br />You are receiving this because you are subscribed to this thread.<br />Reply to this email directly, <a href="https://github.com/kubernetes/kubernetes/pull/42156#issuecomment-305567214">view it on GitHub</a>, or <a href="https://github.com/notifications/unsubscribe-auth/AAq2ClcqsT3Uh8hm762xNI3VkyZBBsBuks5r_vgogaJpZM4MNExi">mute the thread</a>.<img alt="" height="1" src="https://github.com/notifications/beacon/AAq2Chf22c5oAq9vEiWsbLMgFOcij0tLks5r_vgogaJpZM4MNExi.gif" width="1" /></p>
<div itemscope itemtype="http://schema.org/EmailMessage">
<div itemprop="action" itemscope itemtype="http://schema.org/ViewAction">
  <link itemprop="url" href="https://github.com/kubernetes/kubernetes/pull/42156#issuecomment-305567214"></link>
  <meta itemprop="name" content="View Pull Request"></meta>
</div>
<meta itemprop="description" content="View this Pull Request on GitHub"></meta>
</div>

<script type="application/json" data-scope="inboxmarkup">{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/kubernetes/kubernetes","title":"kubernetes/kubernetes","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/kubernetes/kubernetes"}},"updates":{"snippets":[{"icon":"PERSON","message":"@k8s-merge-robot in #42156: @croomes PR needs rebase"}],"action":{"name":"View Pull Request","url":"https://github.com/kubernetes/kubernetes/pull/42156#issuecomment-305567214"}}}</script>
----==_mimepart_593052288ca2_14113f9bb3f3dc3c1098a9--

croomes commented:
thanks @msau42 - rebasing now

saad-ali commented:
https://github.com/kubernetes/kubernetes/pull/46076 just merged. Since this volume has a `SecretRef`, make sure to add this volume to `VisitPVSecretNames(...)` in https://github.com/kubernetes/kubernetes/blob/master/pkg/api/persistentvolume/util.go or else the node may not have access to the secret.

saad-ali commented on pkg/volume/storageos/storageos.go:
```diff
@@ -0,0 +1,574 @@
+	if err != nil {
+		return "", err
+	}
+	return volumeSource.VolumeName, nil
```
You should return storageOS namespace/volumename here instead since they are more unique than `volumesource.VolumeName`.

croomes replied:
done

saad-ali commented on pkg/volume/storageos/storageos.go:
```diff
@@ -0,0 +1,574 @@
+	}
+
+	// Attach the StorageOS volume as a block device
+	devicePath, err := b.manager.AttachVolume(b)
```
Why is the attach happening inside the mount method?

saad-ali replied:
It looks like this `manager.AttachVolume` just sets up a global mount for the device, it does not actually attach the device? If that is the case, and StorageOS has no equivalent of calling the StorageOS control plane and making a volume available on a particular node, then this is fine.

croomes replied:
yes - we have a global namespace so all volumes are visible on all nodes without attaching, though only a single node can mount at a time.  The AttachVolume here does some setup if the volume's device can't be used directly.

liggitt replied:
> It looks like this manager.AttachVolume just sets up a global mount for the device, it does not actually attach the device? If that is the case, and StorageOS has no equivalent of calling the StorageOS control plane and making a volume available on a particular node, then this is fine.

if that's the case, can we rename the function? Mount calling Attach is really confusing, given the rest of the k8s storage interactions

saad-ali commented on staging/src/k8s.io/client-go/pkg/api/types.go:
```diff
@@ -1149,6 +1155,36 @@ type ScaleIOVolumeSource struct {
+	// storage resources used in them.  If not specified, the default pool will be
+	// used.
+	// +optional
+	Pool string
```
Why is this part of the API? If it is only used for provisioning, then it can be an opaque parameter set via the `StorageClass` for provisioning. The information here should only be "connection information" required to mount/attach a provisioned volume.

croomes replied:
I can remove this, but again it seems quite common to have the storage pool part of the API (Gluster, RBD, ScaleIO).

msau42 replied:
I took a quick look at the current plugins:
* gluster doesn't have storage pool in its API
* rbd does have pool in its API, but it needs it for mounting
* scaleio does have it and it seems like its only used for provisioning, so we missed that one

Saad's main point is that if the storage pool is only used for provisioning, and not used for mounting or attaching, then it should just be a parameter in the StorageClass and not in the volume source API.

croomes replied:
ok, understood.  I've removed and am about to re-push.

msau42 replied:
Can you also address this in your update?  It's really hard to change the API once it's in.

saad-ali commented on pkg/volume/storageos/storageos.go:
```diff
@@ -0,0 +1,574 @@
+		if pv.Labels == nil {
+			pv.Labels = make(map[string]string)
+		}
+		for k, v := range vol.Labels {
```
The copying of all labels from PVC to PV object is unprecedented. I would prefer if you removed this unless there is strong justification for it.

croomes replied:
I believe the AWS, GCE, Flocker and Portworx provisioners also do this.  We think it's a useful feature as it allows k8s tools to query/verify storage parameters directly without having to use our api and tools.  

For example, you could query on PVs that have the production label where storageos.feature.replicas < 1 to show data at risk.

saad-ali replied:
I don't feel super strongly about this. Fine to leave it in.

saad-ali reviewed (changes requested):
Some more questions/comments. Code freeze is 6 PM PT today. If this PR gets LGTM/Approve before then we're good.

liggitt commented on pkg/volume/storageos/storageos_util.go:
```diff
@@ -0,0 +1,398 @@
+		apiVersion: defaultAPIVersion,
+	}
+
+	keys, err := a.host.GetKubeClient().Core().Secrets(a.namespace).Get(apiSecretName, metav1.GetOptions{})
```
where is this namespace coming from?

saad-ali replied:
To add to this. When fetching a secret for a volume, the namespace used should always be the same as the pod and not user configurable -- this is to prevent a user in one namespace from being able to steal a secret from another namespace. Because of this you may want to decouple your StorageOS namespaces from k8s namespaces.

liggitt commented on pkg/volume/storageos/storageos_util.go:
```diff
@@ -0,0 +1,398 @@
+		return c
+	}
+	if err != nil {
+		keys, err = a.host.GetKubeClient().Core().Secrets("default").Get(apiSecretName, metav1.GetOptions{})
```
falling back to the default namespace doesn't seem right... is there precedent for this?

liggitt commented on pkg/volume/storageos/storageos_util.go:
```diff
@@ -0,0 +1,398 @@
+		if err != nil {
+			return c
+		}
+	}
```
all other storage-class-based secret lookups require the secret type to match the provisioner name

liggitt commented on examples/volumes/storageos/README.md:
```diff
@@ -0,0 +1,466 @@
+kind: Secret
+metadata:
+  name: storageos-api
+type: Opaque
```
the type should match the provisioner name (see other dynamic provisioners for an example)

liggitt commented on examples/volumes/storageos/README.md:
```diff
@@ -0,0 +1,466 @@
+
+The StorageOS provider has been pre-configured to use the StorageOS API defaults, and no additional configuration is required for testing.  If you have changed the API port, or have removed the default account or changed its password (recommended), you must specify the new settings.  This is done using Kubernetes [Secrets](../../../docs/user-guide/secrets/).
+
+To add a secret to provide custom API configuration, you must name the secret `storageos-api`.  Normally you would add the secret in the `default` namespace so it would be available to all namespaces to use.  You may also add it to a specific namespace for only that namespace to use, allowing different namespaces to authenticate to the API with different user accounts.
```
pinning the name is unusual... it's more typical to specify the secret name and namespace

liggitt commented on examples/volumes/storageos/README.md:
```diff
@@ -0,0 +1,466 @@
+
+## API Configuration
+
+The StorageOS provider has been pre-configured to use the StorageOS API defaults, and no additional configuration is required for testing.  If you have changed the API port, or have removed the default account or changed its password (recommended), you must specify the new settings.  This is done using Kubernetes [Secrets](../../../docs/user-guide/secrets/).
```
is this secret used at provision time, attach/detach time, or mount/unmount time? if mount/unmount, it needs to be represented in the pod volume spec (just as a secret name, current namespace is implicit) and in the pv spec (as a secret name and namespace)... volume plugin mount operations on the node don't make use of information in storageclasses.

croomes commented:
Updating for the secrets changes and GetVolumeName() changes now.

saad-ali commented:
/approve

I, @msau42, or @liggitt will add lgtm once feedback is addressed.

croomes commented:
Almost done - the secrets changes taking longer than expected.  Think I need another half hour or so.

croomes commented:
Review comments should now all be addressed.  Testing against the default StorageOS config works and I'm continuing to test the changes to secrets.  Hopefully this can still make it and any other needed changes can be done as a separate PR?

saad-ali commented on pkg/volume/storageos/storageos.go:
```diff
@@ -0,0 +1,599 @@
+			namespace:       pod.Namespace,
+			pvName:          spec.Name(),
+			volName:         source.VolumeName,
+			volNamespace:    source.VolumeNamespace,
```
The problem @liggitt pointed out still exists: the `VolumeNamespace` is used eventually to fetch secrets: `kubeClient.Core().Secrets(a.namespace).Get(...)`. This means that a user could use this code to access secrets in another namespace. Please make sure that anywhere secrets are accessed the namespace is always set to the namespace of the pod and not something the user can set.

croomes replied:
Thanks - looking at this now + will complete the exception.  Also missed VisitPodSecretNames().

saad-ali reviewed (changes requested):
Code freeze for 1.7 went into effect at 6 PM PT. So technically this PR missed the boat.

However, since it is so close to being complete and very low risk to the rest of k8s code, let's file for exception. @croomes the exception process is documented [here](https://github.com/kubernetes/features/blob/master/EXCEPTIONS.md). Please complete that ASAP. As sig-storage lead I will support this request. The release czar for 1.7 is @dchen1107.

liggitt commented:
one thing that would make review much easier would be to split generated/non-generated changes into separate commits

liggitt commented on staging/src/k8s.io/client-go/pkg/apis/extensions/types.go:
```diff
@@ -926,6 +926,7 @@ var (
 	Quobyte               FSType = "quobyte"
 	AzureDisk             FSType = "azureDisk"
 	PhotonPersistentDisk  FSType = "photonPersistentDisk"
+	StorageOS             FSType = "storageos"
```
`storageos` or `storageOS`?

liggitt commented on pkg/apis/extensions/types.go:
```diff
@@ -926,6 +926,7 @@ var (
 	Quobyte               FSType = "quobyte"
 	AzureDisk             FSType = "azureDisk"
 	PhotonPersistentDisk  FSType = "photonPersistentDisk"
+	StorageOS             FSType = "storageos"
```
`storageOS`?

caesarxuchao replied:
Please copy the const to pkg/apis/extensions/v1beta1/types.go as well.

caesarxuchao replied:
It looks like the list in [v1beta1/types.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/extensions/v1beta1/types.go#L940) has already diverged, please keep them consistent if possible.

Also, is it intentional to make these variables instead of constants?

liggitt commented on pkg/volume/storageos/storageos.go:
```diff
@@ -0,0 +1,599 @@
+// storageos volumes represent a bare host directory mount of an StorageOS export.
+type storageos struct {
+	podUID       types.UID
+	namespace    string
```
rename to podNamespace?

liggitt commented on pkg/volume/storageos/storageos.go:
```diff
@@ -0,0 +1,599 @@
+
+	// Set from PVC
+	c.volName = c.options.PVName
+	c.volNamespace = c.options.PVC.Namespace
```
is this right? I thought volNamespace was the storageOS namespace, not the k8s namespace

liggitt commented on pkg/volume/storageos/storageos.go:
```diff
@@ -0,0 +1,599 @@
+				v1.ResourceName(v1.ResourceStorage): resource.MustParse(fmt.Sprintf("%dGi", vol.Size)),
+			},
+			PersistentVolumeSource: v1.PersistentVolumeSource{
+				StorageOS: &v1.StorageOSVolumeSource{
```
does this mean dynamically provisioned storageOS PVs don't support secret references (and therefore don't support storageOS installations that have removed or changed the default username/password)?

liggitt commented on pkg/volume/storageos/storageos_util.go:
```diff
@@ -0,0 +1,433 @@
+// storageosVolume describes a provisioned volume
+type storageosVolume struct {
+	ID          string
+	Name        string
```
does this correspond to the name of a kubernetes object?

liggitt commented on pkg/volume/storageos/storageos_util.go:
```diff
@@ -0,0 +1,433 @@
+type storageosVolume struct {
+	ID          string
+	Name        string
+	Namespace   string
```
does this correspond to the name of a kubernetes namespace? document it either way, and possibly rename if this is a StorageOS namespace

croomes commented:
@saad-ali @liggitt - thanks for all the feedback and help.  I've sent the exception request asking for 2 more days which should be enough time to clear up the remaining issues and do some more comprehensive testing.   It's pretty late in the UK so I'm going to start fresh tomorrow.

saad-ali commented:
> @saad-ali @liggitt - thanks for all the feedback and help. I've sent the exception request asking for 2 more days which should be enough time to clear up the remaining issues and do some more comprehensive testing. It's pretty late in the UK so I'm going to start fresh tomorrow.

Thanks for all your hard work!

liggitt commented on pkg/volume/storageos/storageos_util.go:
```diff
@@ -0,0 +1,433 @@
+		return nil, fmt.Errorf("Cannot get kube client")
+	}
+
+	keys, err := kubeClient.Core().Secrets(a.namespace).Get(a.secretName, metav1.GetOptions{})
```
the pkg/volume/util#GetSecretForPV and GetSecretForPod helper functions enforce best practices (pulls namespace from Pod when fetching for a pod-referenced volume source, and ensures secret type matches when fetching for a PV-referenced volume source. Check how other plugins call those to see how they plumb the data from the source.

liggitt commented on pkg/volume/storageos/storageos_util.go:
```diff
@@ -0,0 +1,433 @@
+		},
+		Type: "kubernetes.io/storageos",
+	}
+	_, err := kubeClient.Core().Secrets(namespace).Create(secret)
```
this is unexpected... how is this used, and what component is expecting to have permission to create secrets in arbitrary namespaces?

liggitt commented on pkg/volume/storageos/storageos.go:
```diff
@@ -0,0 +1,599 @@
+	defaultNamespace    = "default"
+)
+
+func getPath(uid types.UID, namespace string, volName string, pvName string, host volume.VolumeHost) string {
```
a general comment throughout would be to clarify what type of namespace this is... storageos namespace or kubernetes namespace. I'm concerned the former (which is under the user's control in the volume source?) will be used as the other and allow cross-namespace access of some type

croomes commented:
Quick update: we made quite a few changes (separate commit) that hopefully are using the secrets better and clearing up the namespace usage.  I still need to change storageos.namespace -> storageos.podNamespace so there are no ambiguities left.

I'm confused how to get the secret data for the unmount operation.  I think I need to create a new secret for each volume (called namespace/volName) with the secretName as data during Mount, then fetch it during Unmount.  Does that sound right or is there a better way?

Our tests also need updating - I expect to have the code completed and tested on Monday.

croomes commented:
@k8s-bot pull-kubernetes-unit test this
@k8s-bot pull-kubernetes-bazel test this
@k8s-bot pull-kubernetes-e2e-kops-aws test this

croomes commented:
@saad-ali / @liggitt - I've updated to hopefully address the secret namespace and other issues.  It took a bit of a refactor so a more changes than I expected.  I haven't heard back about the extension but if you could have a look that would be great.

saad-ali commented:
> @saad-ali / @liggitt - I've updated to hopefully address the secret namespace and other issues. It took a bit of a refactor so a more changes than I expected. I haven't heard back about the extension but if you could have a look that would be great.

Taking a look

saad-ali commented on pkg/api/types.go:
```diff
@@ -1149,6 +1155,34 @@ type ScaleIOVolumeSource struct {
+	// SecretName specifies the secret to use for obtaining the StorageOS API
+	// credentials.  If not specified, default values will be attempted.
+	// +optional
+	SecretName string
```
Why a string? Should do `SecretRef *LocalObjectReference`

saad-ali reviewed (changes requested):
> I'm confused how to get the secret data for the unmount operation. I think I need to create a new secret for each volume (called namespace/volName) with the secretName as data during Mount, then fetch it during Unmount. Does that sound right or is there a better way?

The creation of secrets by a volume plugin is unexpected and should be avoided. It looks like you need access to 1. the pod namespace and 2. the secret name at mount/unmount time. In order to do this, your volume source will contain the secret name (`SecretRef`), and you can save the pod namespace in `NewMounter` and use it in `Setup(...)`

liggitt commented on pkg/api/types.go:
```diff
@@ -1149,6 +1155,34 @@ type ScaleIOVolumeSource struct {
+	// Set to "default" if you are not using namespaces within StorageOS.
+	// Namespaces that do not pre-exist within StorageOS will be created.
+	// +optional
+	VolumeNamespace string
```
this still seems like it allows cross-kubernetes-namespace mounting of the same storageos volume. is that intended?

liggitt commented on pkg/api/validation/validation.go:
```diff
@@ -1116,6 +1124,14 @@ func validateLocalVolumeSource(ls *api.LocalVolumeSource, fldPath *field.Path) f
+	allErrs := field.ErrorList{}
+	if len(storageos.VolumeName) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("volumeName"), ""))
+	}
```
are there any format restrictions on the VolumeNamespace?

croomes replied:
yes, good point.  Will add.  (We use the same Label format as kube).

liggitt commented on pkg/volume/storageos/storageos.go:
```diff
@@ -0,0 +1,786 @@
+)
+
+func getPath(uid types.UID, volNamespace string, volName string, pvName string, host volume.VolumeHost) string {
+	if len(volNamespace) != 0 && len(volName) != 0 && strings.Count(volName, ".") == 0 {
```
what are the two scenarios being differentiated between here? which parameters are under the user's control?

croomes replied:
The volNamespace and volName are supplied by the user.  This goes back to my issue with the unmount; from what I can tell the unmount only gets the podUID and pvName, but we need the volName and volNamespace (and API credentials) to reference the volume in our API to remove the mount reference.  Encoding them in the mountpath was a way to get the volNamespace & volName.

Unless there's another way, since we're not able to get the creds I could remove this too.

liggitt commented on pkg/volume/storageos/storageos.go:
```diff
@@ -0,0 +1,786 @@
+	var secretNamespace, secretName string
+
+	// SecretName will be a locator ref for dynamically provisioned volumes
+	if strings.HasPrefix(source.SecretName, secretLocatorPrefix) {
```
this doesn't seem like a reliable way to plumb this info, especially since it means the secret can be fetched from a namespace other than the pod namespace

liggitt commented on pkg/volume/storageos/storageos.go:
```diff
@@ -0,0 +1,786 @@
+func (plugin *storageosPlugin) NewProvisioner(options volume.VolumeOptions) (volume.Provisioner, error) {
+	var secretName string
+	for k, v := range options.Parameters {
+		if strings.ToLower(k) == "secretname" {
```
are these from a storageclass? if so, why not specify a secret namespace as a separate parameter? (c.f. https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/quobyte/quobyte.go#L448)

liggitt commented on pkg/volume/storageos/storageos.go:
```diff
@@ -0,0 +1,786 @@
+	c.size = int(volume.RoundUpSize(capacity.Value(), 1024*1024*1024))
+
+	// Store API secret location for lookup during unmount/delete.
+	secretRef, err := c.plugin.storeSecretRef(c.podNamespace, c.secretName, c.options.PVC.UID)
```
creating secrets in every namespace isn't appropriate... if you're going to denormalize, I'd rather see the secret name and namespace specified in the PV spec (though that means reworking https://github.com/kubernetes/kubernetes/blob/master/pkg/api/persistentvolume/util.go#L26 to pass secret name and namespace to the visitor, since the storageos volume source would differ from the other sources)

liggitt commented:
if I'm reading this correctly, there are three places secrets can be referenced from:
1. storage class options (should be able to specify secret namespace here)
2. persistent volume source
3. pod spec volume source (should always use pod namespace)

I'd like to see the secret name/namespace determination made directly from the source of the information, rather than having the namespace defaulted, or plumbed as a prefix of the secret name depending on the source type.

Are the credentials required at mount time, on the node? Is it possible to attach the storage to the node externally? That is preferred to avoid exposing credentials to nodes, especially in the case where all PVs that are auto-provisioned from a storage class will share a single credential... exposing that to nodes isn't ideal.

saad-ali commented:
@croomes If it would be helpful, feel free to set up a video conference with @liggitt and I. I am free until 1:30 PM PT today and after 4:30 PM PT.

croomes commented:
@saad-ali thanks, but I've only just seen your msg - that would have been really helpful.  I've made the changes to the SecretRef and removed creating the secret so we can locate the real secret on unmount.

I'm still stuck is on Unmount, where I can't figure out a way to get the API credentials.  I changed our API util to release any existing internal mount reference before calling our API to mount as we protect against multiple mounts internally.  Everything works in this PR update, but it show volumes as mounted (within our control plane only) after it has been unmounted in kube.

The only way I can think of getting around this is to encode the secret name in the mountpath or write a config file, but both are really ugly and I know won't be popular.  So unless there's another way, we're happy to leave it like this since Kube is protecting against concurrent mounts itself and we can explain in our docs.

So, would really appreciate another look from you and @liggitt when you get a chance.  I'll be available at ~4:30 PT and be keeping a close eye for a couple hours.

croomes commented:
fixing test errors now - missed an update

croomes commented:
Just to reply directly to some questions:

> Are the credentials required at mount time, on the node? 

yes, on mount we make an API call to mark the volume as mounted so it can't be re-mounted concurrently elsewhere.

> Is it possible to attach the storage to the node externally? That is preferred to avoid exposing credentials to nodes, especially in the case where all PVs that are auto-provisioned from a storage class will share a single credential... exposing that to nodes isn't ideal.

No credentials are needed to attach the volume - every volume will be visible as a block device on all nodes running the StorageOS container.  Mounting requires API credentials and permission to that volume.

saad-ali commented:
> So, would really appreciate another look from you and @liggitt when you get a chance. I'll be available at ~4:30 PT and be keeping a close eye for a couple hours.

Thanks. Taking a look. We can set up a meeting at 4:30 PM PT we have any outstanding issues after this round.

saad-ali commented on pkg/volume/storageos/storageos.go:
```diff
@@ -0,0 +1,705 @@
+
+// secretRefName returns a prefixed name based on the pod or PV UID so that it
+// will be unique within the default namespace.
+func secretRefName(UID types.UID) string {
```
Doesn't look like this is used anywhere. Can be removed along with `secretLocatorPrefix`?

croomes replied:
yes - good catch

saad-ali commented on pkg/volume/storageos/storageos.go:
```diff
@@ -0,0 +1,705 @@
+		switch strings.ToLower(k) {
+		case "adminsecretname":
+			adminSecretName = v
+		case "adminsecretnamespace":
```
Do not allow user to control `secretNamespace` for provisioning either (remove this parameter). Infer the secret namespace for provisioning from the PVC (i.e. use the PVC namespace).

croomes replied:
ahh.. yes

liggitt replied:
This is specified in the storage class, right? That's a privileged object to write, and other plugins let you fully specify the coordinates of a secret in storage class parameters

liggitt replied:
If you require a secret for a dynamically provisioned PV to be in the PVC, you either have to rely on the user providing it (unlikely), or you have to make the provisioner place a secret in the namespace (not good)

croomes replied:
Yes, it's in the storage class currently.

The idea behind allowing it to be set was to support the secret being in the PVCs namespace and also the default namespace (I bet many installs will have a single admin credential).

Would it work if I add a check to enforce it matches either PVC namespace or default?

saad-ali replied:
> This is specified in the storage class, right? That's a privileged object to write, and other plugins let you fully specify the coordinates of a secret in storage class parameters
> 
> If you require a secret for a dynamically provisioned PV to be in the namespace of the PVC, you either have to rely on the user providing it (unlikely), or you have to make the provisioner place a secret in the namespace (not good)

@liggitt Good point. I would be fine leaving it as is then, if you're ok with it.

> Would it work if I add a check to enforce it matches either PVC namespace or default?

@croomes At that point might as well remove it as a parameter. So either keep it as is (if @liggitt is ok with that), or remove the `adminsecretnamespace` parameter altogether and use the PVC's namespace. I'm ok with either.

croomes replied:
My preference is to leave it.  I'm about to push the other changes, but can remove later it if needed.

saad-ali reviewed (changes requested):
Almost there. You are letting user specify secret via `SecretRef` but getting secret namespace from pod which is good. However, for provisioning you are allowing user to pass in both `adminsecret` and `adminsecretnamespace` -- let's fix that.

croomes commented:
Should I squash to a single commit before I push?  update-all almost finished.

I've added ValidateDNS1123Label to VolumeName and VolumeNamespace in pkg/api/validation/validation.go, and removed the orphaned var and function in storageos.go.

I don't think there is anything else outstanding?

saad-ali commented:
Go for it. I don't have any more feedback. If @liggitt is ok with it, I'll LGTM.

liggitt commented on pkg/volume/storageos/storageos.go:
```diff
@@ -0,0 +1,698 @@
+		return nil, err
+	}
+	adminSecretName := ""
+	adminSecretNamespace := "default"
```
I would not expect a fallback to a default namespace. Either specify the namespace in the storage class, or use the namespace of the PVC, or don't fetch a secret

liggitt replied:
hmm, I see quobyte and rbd are doing this... that is unfortunate.

liggitt replied:
this still needs to be removed

liggitt commented on pkg/volume/storageos/storageos.go:
```diff
@@ -0,0 +1,698 @@
+					FSType:          vol.FSType,
+					ReadOnly:        false,
+					SecretRef: &v1.LocalObjectReference{
+						Name: adminSecretName,
```
The other volume plugins that make use of a secret name/namespace in the storage class parameters only do so during the Provision()/Delete() operations, which are run by the persistent volume controller, which can look up the storage class and any secret.

This plugin expects to have access to that secret from the node as part of the mount/unmount operation.

The adminSecretNamespace specified in the storage class is lost here...
1. that means the node mount/unmount operation won't know what namespace to look in to find the secret referenced by the storage class
2. even if it knew where to look, it would be forbidden (when running with the Node authorizer enabled)

For the node to have the info required to locate the secret, and the authorizer to allow the node to access the secret, I think the PV volume source has to allow specifying the secret name and namespace (which means the PV storageos volume source looks different than the Pod storageos volume source which only allows a local reference).

Then the mount/unmount operation has the info needed to locate the secret, and the node authorizer has the info needed to permit access to it.

croomes replied:
Yes, I see the issue now.  To confirm, I should change the LocalObjectReference to an ObjectReference with Namespace set to adminSecretNamespace, then update the Visitor functions to add SecretRef.Namespace?

This is simple enough to do quickly now but if the extension can hold until tomorrow I'd like to re-submit after adding some more thorough acceptance tests in the morning.  Is that ok?

liggitt replied:
> To confirm, I should change the LocalObjectReference to an ObjectReference with Namespace set to adminSecretNamespace, then update the Visitor functions to add SecretRef.Namespace?

That would give the node running the mount/unmount operation for the PV the info it needs, and the node authorizer the info it needs, so it makes sense to me.
* cc @kubernetes/sig-storage-api-reviews on the data flow for the storage plugin
* cc @kubernetes/sig-auth-api-reviews on the data flow for the node authorizer

I would expect a storage class to specify both secret name and namespace, or neither (to use default storageos credentials). If just the secret name or namespace, I'd expect the dynamic provision to fail, not fall back to looking in the `default` namespace.

Also, this is the first volume source I know of that differs between the PV spec and the Pod spec (pod spec would only allow a local secret reference, pv spec would allow a fully specified name/namespace reference). That makes sense to me, given the use, but I wanted to call it out.

liggitt replied:
> This is simple enough to do quickly now but if the extension can hold until tomorrow I'd like to re-submit after adding some more thorough acceptance tests in the morning. Is that ok?

Outcome of extension requests will be determined during the 1.7 burndown meeting tomorrow (Wednesday) afternoon

croomes replied:
Added this and did a lot of testing, it works well.  Running update-all and will push soon.

liggitt commented on pkg/volume/storageos/storageos.go:
```diff
@@ -0,0 +1,698 @@
+	description  string
+	pool         string
+	fsType       string
+	size         int
```
doc what size is. bytes? GB?

liggitt replied:
I see some other volume plugins name the field as sizeGB... probably a good idea

croomes replied:
done in next push

liggitt commented on pkg/volume/storageos/storageos_util.go:
```diff
@@ -0,0 +1,374 @@
+	Namespace   string
+	Description string
+	Pool        string
+	Size        int
```
doc size units

croomes replied:
done in next push

liggitt commented on pkg/api/persistentvolume/util.go:
```diff
@@ -50,6 +50,10 @@ func VisitPVSecretNames(pv *api.PersistentVolume, visitor func(string) bool) boo
 			return false
 		}
+	case source.StorageOS != nil:
+		if source.StorageOS.SecretRef != nil && !visitor(source.StorageOS.SecretRef.Name) && !visitor(source.StorageOS.SecretRef.Namespace) {
```
the signature of the visitor has to change to take (namespace, name)

croomes replied:
Do I need both ways: visitor(name) and if namespace set, visitor(namespace, name)?

liggitt replied:
right now, the sole use of VisitPVSecretNames is inferring secret namespace from the PV's `pv.Spec.ClaimRef.Namespace`, because that's how all secret references from a PV were resolved until now.

With this change, the storageos secret namespace is separately specified.

For all the other calls to the visitor, pass `pv.Spec.ClaimRef.Namespace` as the secret namespace (and if pv.Spec.ClaimRef is nil, don't call the visitor)

liggitt commented on pkg/api/v1/types.go:
```diff
@@ -1228,6 +1234,34 @@ type ScaleIOVolumeSource struct {
+	// SecretRef specifies the secret to use for obtaining the StorageOS API
+	// credentials.  If not specified, default values will be attempted.
+	// +optional
+	SecretRef *ObjectReference `json:"secretRef,omitempty" protobuf:"bytes,5,opt,name=secretRef"`
```
when used within a podspec, this doesn't make sense... this is what I was saying about needing on StorageOSVolumeSource struct within a pod spec which used a LocalObjectReference for the secret, and an almost identical StorageOSPersistentVolumeSource struct within a PV spec which used an ObjectReference

croomes replied:
I think I understand now: In types.go I create a StorageOSPersistentVolumeSource (1st plugin to do this) and use it in Provision() when returning the PV spec?

liggitt replied:
right, it's a very slightly different type than is used in a pod spec volume. It has ripples into the volume plugin, since it means there are two flavors of storageos volume sources :-/

that's one reason I wanted to get an opinion from @kubernetes/sig-storage-api-reviews on this

liggitt commented on pkg/api/validation/validation.go:
```diff
@@ -1116,6 +1124,20 @@ func validateLocalVolumeSource(ls *api.LocalVolumeSource, fldPath *field.Path) f
 	return allErrs
 }
 
+func validateStorageOSVolumeSource(storageos *api.StorageOSVolumeSource, fldPath *field.Path) field.ErrorList {
```
when validating a PV storageos volume source, need to make sure that the secret name and namespace are both specified, or neither is specified, right?

croomes commented:
Last update removes setting secret namespace to default, and adds validation for name/namespace.  Looking at StorageOSPersistentVolumeSource and visitor function now.

croomes commented:
@liggitt I've just re-pushed, would appreciate if you can confirm I'm on the right track while I re-test:
- added StorageOSPersistentVolumeSource to types.go
- return it in Provision()
- added namespace to visitor in pkg/api/persistentvolume/util.go.  I think it's ok but very unsure about it.

liggitt commented on plugin/pkg/auth/authorizer/node/graph.go:
```diff
@@ -251,7 +251,7 @@ func (g *Graph) AddPV(pv *api.PersistentVolume) {
-		pvutil.VisitPVSecretNames(pv, func(secret string) bool {
+		pvutil.VisitPVSecretNames(pv, func(namespace, secret string) bool {
 			// This grants access to the named secret in the same namespace as the bound PVC
 			g.graph.SetEdge(simple.Edge{F: g.getOrCreateVertex_locked(secretVertexType, pv.Spec.ClaimRef.Namespace, secret), T: pvVertex})
```
use the provided namespace here instead of `pv.Spec.ClaimRef.Namespace`

croomes replied:
done

liggitt commented on pkg/api/persistentvolume/util_test.go:
```diff
@@ -54,11 +55,22 @@ func TestPVSecrets(t *testing.T) {
 	for _, pv := range pvs {
-		VisitPVSecretNames(pv, func(name string) bool {
+		VisitPVSecretNames(pv, func(namespace, name string) bool {
+			fmt.Printf("%s,%s\n", namespace, name)
```
drop the printf

croomes replied:
whoops - done

liggitt commented on pkg/volume/storageos/storageos.go:
```diff
@@ -0,0 +1,745 @@
+
+func (plugin *storageosPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, _ volume.VolumeOptions) (volume.Mounter, error) {
+
+	source, _, err := getPersistentVolumeSource(spec)
```
doesn't this mean the storageos plugin only supports persistentvolumes, not a storageos volume source specified in a pod spec?

croomes replied:
No - this is a mistake.  I added the code to detect both in newMounterInternal but not NewMounter (my tests were failing on it).  I will refactor a bit and add to both.  I couldn't think of a better way to handle both cases.

liggitt replied:
yeah, this is part of the ripple of not being able to share a single volume source :-/

liggitt commented on pkg/volume/storageos/storageos.go:
```diff
@@ -0,0 +1,745 @@
+	}
+
+	var apiCfg *storageosAPIConfig
+	if source.SecretRef != nil {
```
I expected two paths here, one using secret name/namespace for a PV source, one using secret name and pod namespace for a pod source

saad-ali commented:
@dchen1107 held the v1.7 exceptions meeting today. She left it up to @liggitt and I to decide if this PR gets in for 1.7 or not.

@liggitt and I just had a meeting to discuss the current revision of this PR where the VolumeSource for this plugin is split between a PV (with `ObjectReference`) and non-PV (with `LocalObjectReference`) VolumeSource. We agree that while this is different from existing plugins, it is necessary to handle the unique access pattern for StorageOS--where they have a global credential that is required for provisioning/mounting/unmounting/deleting. For the dynamically provisioned volumes both secret namespace and name are specified, but since the PV object is non-namespaced, it can only be created with some one with admin privileges so that is ok. In this case the secret can live in a non-user namespace and therefore even if the user deletes their entire namespace, the volume will still be deleted correctly. For direct volume references, the secret must live in the users namespace (deletion is not a worry).

So overall we are fine with this approach. @liggitt and I will both take a final pass reviewing this PR. In the meantime, @croomes please test all the paths for this code (provisioning, mounting, unmounting, deleting, etc.). Once we have a thumbs up from you on the testing front and no further feedback, we will approve/LGTM this for merge (Friday at the latest).

croomes commented:
Fantastic, thanks!  I'm going to do another push in 30 mins that should fix the last comments, but I will spend the day tomorrow double-checking everything.  I've re-run our basic acceptance tests and they pass for PVs and volumes in Pod specs and dynamic provisioning.  I need to do more with the various secret namespace combinations. 

To help review latest, I've only changed NewMounter and newMounterInternal, and changed validation.go to return proper path to secretRef name/namespace.  I couldn't find another way except to set fldPath.Child("secretRef.namespace"), which is unusual.  I now get:

> The PersistentVolume "pv0001" is invalid: spec.storageos.secretRef.namespace: Required value

Where on just fldPath.Child("namespace") it was reporting "spec.storageos.namespace"

Really appreciate all the time you've both put into helping me get this in. :+1:

liggitt commented on pkg/api/persistentvolume/util.go:
```diff
@@ -20,34 +20,46 @@ import (
 	"k8s.io/kubernetes/pkg/api"
 )
 
+func getNamespace(pv *api.PersistentVolume) string {
```
nit: rename this to `getClaimRefNamespace`

liggitt commented on pkg/api/validation/validation.go:
```diff
@@ -1116,6 +1124,40 @@ func validateLocalVolumeSource(ls *api.LocalVolumeSource, fldPath *field.Path) f
+	if len(storageos.VolumeNamespace) > 0 {
+		allErrs = append(allErrs, ValidateDNS1123Label(storageos.VolumeNamespace, fldPath.Child("volumeNamespace"))...)
+	}
+	return allErrs
```
if secretRef != nil, ensure a name is set?

liggitt commented on pkg/api/validation/validation.go:
```diff
@@ -1116,6 +1124,40 @@ func validateLocalVolumeSource(ls *api.LocalVolumeSource, fldPath *field.Path) f
+	if len(storageos.VolumeNamespace) > 0 {
+		allErrs = append(allErrs, ValidateDNS1123Label(storageos.VolumeNamespace, fldPath.Child("volumeNamespace"))...)
+	}
+	if storageos.SecretRef != nil && (len(storageos.SecretRef.Name) > 0 || len(storageos.SecretRef.Namespace) > 0) {
```
if secretRef != nil, I'd expect the name and namespace to be set... a non-nil secretRef with empty data seems wrong

liggitt commented on pkg/volume/storageos/storageos.go:
```diff
@@ -0,0 +1,761 @@
+	capacity := c.options.PVC.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)]
+	c.sizeGB = int(volume.RoundUpSize(capacity.Value(), 1024*1024*1024))
+
+	if adminSecretName == "" {
```
actually, does requiring this mean you can no longer use default storageos credentials?

liggitt commented:
one comment about combining getSecretInfoFromSpec and parseSecret to make it very clear that parsePVSecret() is only called with data from a PV, and a couple validation questions.

the graph and pv secret util changes LGTM

croomes commented:
I've made the changes suggested by @liggitt (good ideas), and we've done a lot of testing.  Functionally we're happy, but I want to make one final pass on the examples and README to make sure we didn't miss anything.  I'll have it rebased and ready to go for start of day (PT) Friday.

croomes commented:
@saad-ali / @liggitt - we're good from our side.  A few minor updates to the docs and we've re-tested everything and it all looks fine.

saad-ali commented on pkg/api/types.go:
```diff
@@ -384,6 +387,10 @@ type PersistentVolumeSource struct {
+	// StorageOS represents a StorageOS volume that is attached to the kubelet's host machine and mounted into the pod
+	// More info: https://releases.k8s.io/HEAD/examples/volumes/storageos/README.md
+	// +optional
+	StorageOS *StorageOSPersistentVolumeSource
```
nit: Maybe add a comment indicating why it is not `StorageOSVolumeSource`.

saad-ali reviewed (approved):
/lgtm
/approve

croomes commented:
👍 It's like another election night - the team is in the pub waiting for the votes to come in.  Big cheer from over here - thanks.
